File: src/gateway/Gateway.REST/Services/ChatService.cs
using QuickChat.Gateway.REST.Models;

namespace QuickChat.Gateway.REST.Services;

public class ChatService(Chat.ChatClient client, ICurrentUserProvider currentUserProvider)
    : IChatService
{
    private readonly Chat.ChatClient client = client;
    private readonly ICurrentUserProvider currentUserProvider = currentUserProvider;

    public async Task<ChatModel> GetChatByCode(string code)
    {
        GetChatByCodeRequest request = new() { Code = code };
        GetChatByCodeReply reply = await client.GetChatByCodeAsync(request);
        return MapChat(reply.Chat);
    }

    public async Task<IList<ChatModel>> GetCurrentUserChats()
    {
        GetUserChatsRequest request =
            new() { UserId = currentUserProvider.GetCurrentUserId().ToString() };
        GetUserChatsReply reply = await client.GetUserChatsAsync(request);
        return [.. reply.Chats.Select(MapChat)];
    }

    public async Task<ChatModel> CreateChat(string name)
    {
        CreateChatRequest request = new() { Name = name };
        CreateChatReply reply = await client.CreateChatAsync(request);
        return MapChat(reply.Chat);
    }

    public async Task<ChatParticipantModel> JoinChat(Guid chatId, string name)
    {
        JoinChatRequest request =
            new()
            {
                ChatId = chatId.ToString(),
                Name = name,
                UserId = currentUserProvider.GetCurrentUserId().ToString()
            };
        JoinChatReply reply = await client.JoinChatAsync(request);
        return MapChatParticipant(reply.Participant);
    }

    private static ChatModel MapChat(ChatMessage message)
    {
        return new ChatModel(
            new Guid(message.Id),
            message.Name,
            message.Code,
            [.. message.Participants.Select(MapChatParticipant)],
            message.CreatedAt.ToDateTime(),
            message.LifeTimeSeconds
        );
    }

    private static ChatParticipantModel MapChatParticipant(ChatParticipantMessage message)
    {
        return new ChatParticipantModel(
            new Guid(message.Id),
            new Guid(message.UserId),
            message.Name,
            message.JoinedAt.ToDateTime()
        );
    }
}

File: src/clients/vue/src/services/api/chatService.ts
import type Chat from '@/models/Chat';
import { apiClient } from './client';
import type ChatParticipant from '@/models/ChatParticipant';

export async function getChatByCode(code: string): Promise<Chat> {
  const { data } = await apiClient.get<Chat>(`/chats/${code}`);
  return data;
}

export async function getUserChats(): Promise<Chat[]> {
  const { data } = await apiClient.get<Chat[]>(`/chats`);
  return data;
}

export async function createChat(name: string): Promise<Chat> {
  const formData = new FormData();
  formData.append('name', name);

  const { data } = await apiClient.post<Chat>('/chats', formData);
  return data;
}

export async function joinChat(chatId: string, name: string): Promise<ChatParticipant> {
  const formData = new FormData();
  formData.append('name', name);

  const { data } = await apiClient.post<ChatParticipant>(`/chats/${chatId}/participants`, formData);
  return data;
}
